server/src/main/java/org/apache/druid/server/http/CatalogResource.java

```diff
@@ -0,0 +1,488 @@
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response createTable(
+      TableMetadata table,
+      @QueryParam("ifnew") boolean ifNew,
+      @Context final HttpServletRequest req)
+  {
+    String dbSchema = table.resolveDbSchema();
+    Pair<Response, SchemaSpec> result = validateSchema(dbSchema);
+    if (result.lhs != null) {
+      return result.lhs;
+    }
+    SchemaSpec schema = result.rhs;
+    if (!schema.writable()) {
+      return Actions.badRequest(
+          Actions.INVALID,
+          StringUtils.format("Cannot create tables in schema %s", dbSchema));
+    }
+    table = table.withSchema(dbSchema);
+    try {
+      table.validate();
+    }
+    catch (IAE e) {
+      return Actions.badRequest(Actions.INVALID, e.getMessage());
+    }
+    TableSpec spec = table.spec();
+    if (!schema.accepts(spec)) {
+      return Actions.badRequest(
+          Actions.INVALID,
+          StringUtils.format(
+              "Cannot create tables of type %s in schema %s",
+              spec == null ? "null" : spec.getClass().getSimpleName(),
+              dbSchema));
+    }
+    try {
+      catalog.authorizer().authorizeTable(schema, table.name(), Action.WRITE, req);
+    }
+    catch (ForbiddenException e) {
+      return Actions.forbidden(e);
+    }
+    try {
+      long createVersion = catalog.tables().create(table);
+      return Actions.okWithVersion(createVersion);
+    }
+    catch (DuplicateKeyException e) {
+      if (!ifNew) {
+        return Actions.badRequest(
+              Actions.DUPLICATE_ERROR,
+              StringUtils.format(
+                  "A table of name %s.%s aleady exists",
+                  table.dbSchema(),
+                  table.name()));
+      } else {
+        return Actions.okWithVersion(0);
+      }
+    }
+    catch (Exception e) {
+      return Actions.exception(e);
+    }
+  }
+
+  /**
+   * Update a table within the given schema.
+   *
+   * @param dbSchema The name of the Druid schema, which must be writable
+   *        and the user must have at least read access.
+   * @param name The name of the table definition to modify. The user must
+   *        have write access to the table.
+   * @param spec The new table definition.
+   * @param version An optional table version. If provided, the metadata DB
+   *        entry for the table must be at this exact version or the update
+   *        will fail. (Provides "optimistic locking.") If omitted (that is,
+   *        if zero), then no update conflict change is done.
+   * @param req the HTTP request used for authorization.
+   * @return the new version number of the table
+   */
+  @POST
+  @Path("/tables/{dbSchema}/{name}")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response updateTableDefn(
+      @PathParam("dbSchema") String dbSchema,
+      @PathParam("name") String name,
+      TableSpec spec,
+      @QueryParam("version") long version,
```

Comment:
The version thing is cool. I'm a fan of this sort of thing in CRUD APIs.

Reply:
Killed three birds with one stone: an update time, and a free way to do optimistic locking for those who are into such things. Also helps keep the remote cache in sync. It will be foiled by those who make more than one change per ms, but I suspect that will happen rarely. If it does, the cheat is to sleep for 1ms to bump the number. To be bullet-proof, there needs to be some prevention of moving backwards if the auto clock sync decides our system clock is moving fast and sets it back. We'll fine-tune that later, once the basics are seen to work.